### PR TITLE
Add Chat Recolouring

### PR DESCRIPTION
**Pull request by Hydrox6**
Every type of chat message can now be recoloured. I have done my best to try and make the defaults reflect the actual client defaults.

java_2018-05-08_13-43-23

java_2018-05-08_13-25-18

Some options haven't been tested, such as Broadcasts for the Official Livestream.

Known Issues that are annoying to fix: 
Usernames will not update when the colour is changed or the plugin is disabled.
Some colours won't update when any config is changed while the chatbox is transparent (eg. Clan Chat Info).

Coming soon (unless they should be another PR): Word highlights, and notify on highlight

### CODE REVIEW

**Comment by deathbeam** [comment]
You should remove all the color configurations from other plugins, because they will fight with this one. Also, it would make sense to have all the color configs in 1 plugin.

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java [hunk 1]
```diff
@@ -0,0 +1,589 @@
+package net.runelite.client.plugins.textrecolor;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: Add license headers

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java [hunk 2]
```diff
@@ -0,0 +1,589 @@
+package net.runelite.client.plugins.textrecolor;
+
+import net.runelite.client.config.Config;
+import net.runelite.client.config.ConfigGroup;
+import net.runelite.client.config.ConfigItem;
+
+import java.awt.*;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: Fix your imports according to runelite codestyle on wiki

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+)
+public interface TextRecolorConfig extends Config
+{
+	@ConfigItem(
+		position = 30,
+		keyName = "__dummyTitleOpaque",
+		name = "Opaque Chat Colors",
+		description = ""
+	)
+	default long __dummyTitleOpaque()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: What is this?

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+	{
+		return 0;
+	}
+
+	@ConfigItem(
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: Why are you starting at 31

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
+package net.runelite.client.plugins.textrecolor;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: License headers/fix imports

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+		}
+	}
+
+	@Provides
+	TextRecolorConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(TextRecolorConfig.class);
+	}
+
+	public void checkChatboxState()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: Make this private

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+		{
+			messageBuilder.append(ChatColorType.NORMAL).append(message);
+		}
+
+		messageNode.setRuneLiteFormatMessage(messageBuilder.build());
+		chatMessageManager.update(messageNode);
+		client.refreshChat();
+	}
+
+	private String wrapTextWithColour(String text, Color colour)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: Make this static

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+import javax.inject.Inject;
+import java.awt.*;
+
+@PluginDescriptor(
+	name = "Chat Text Recolor",
+	enabledByDefault = false
+)
+public class TextRecolorPlugin extends Plugin
+{
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: Remove this new line

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java (hunk 3 above)
Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: I'm using config options with invalid types to make an option without a value to use as a header.
For Example: the top value.
java_2018-05-02_08-50-54

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java (hunk 4 above)
Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: I wanted to leave space for settings to go above. I have some code for text highlights that starts at 0. 30 seemed like a good number of spaces to leave.

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java (hunk 3 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorConfig.java]: I would prefer if we didn't do this

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+		}
+
+		final ChatMessageBuilder messageBuilder = new ChatMessageBuilder();
+		final ChatMessageBuilder usernameBuilder = new ChatMessageBuilder();
+		final ChatMessageBuilder clanChannelBuilder = new ChatMessageBuilder();
+
+		MessageNode messageNode = setMessage.getMessageNode();
+
+		//Username Colouring
+		if (setMessage.getName() != "")
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: string compare with !=

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+	@Subscribe
+	public void onSetMessage(SetMessage setMessage)
+	{
+		if (client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		final ChatMessageBuilder messageBuilder = new ChatMessageBuilder();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: I would move these into the respective if's

File: runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+	private boolean isChatboxTransparent = false;
+
+	@Subscribe
+	public void onGameStateChange(GameStateChanged event)
+	{
+		switch (event.getGameState())
+		{
+			case LOGIN_SCREEN:
+			case LOGGING_IN:
+				cacheConfiguredColors();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: missing break?

Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/textrecolor/TextRecolorPlugin.java]: For LOGIN_SCREEN and LOGGING_IN, I also wanted it to refresh the chat manager. I can duplicated the refreshAll() line and break if you think that would read better.

**Comment by MESLewis** [comment]
I suggest that this(or whatever chat color print is working) be cleaned up to support simple recoloring of public, clan, and pm chat. The same config can be used for both transparent and opaque chat. This feature is asked for fairly often and this config situation won't clear up anytime soon.

**Comment by Adam-** [comment]
Added in #3421

**Pull request closed without merging**